packages/x-data-grid-premium/src/hooks/features/pivoting/useGridPivoting.ts

```diff
@@ -0,0 +1,402 @@
+    { rows: GridRowModel[]; columns: Map<string, GridColDef> } | undefined
+  >(undefined);
+
+  const isPivotingAvailable = isPivotingAvailableFn(props);
+
+  apiRef.current.registerControlState({
+    stateId: 'pivotModel',
+    propModel: props.pivotModel,
+    propOnChange: props.onPivotModelChange,
+    stateSelector: gridPivotModelSelector,
+    changeEvent: 'pivotModelChange',
+  });
+
+  apiRef.current.registerControlState({
+    stateId: 'pivotMode',
+    propModel: props.pivotEnabled,
+    propOnChange: props.onPivotEnabledChange,
+    stateSelector: gridPivotEnabledSelector,
+    changeEvent: 'pivotModeChange',
+  });
+
+  apiRef.current.registerControlState({
+    stateId: 'pivotPanelOpen',
+    propModel: props.pivotPanelOpen,
+    propOnChange: props.onPivotPanelOpenChange,
+    stateSelector: gridPivotPanelOpenSelector,
+    changeEvent: 'pivotPanelOpenChange',
+  });
+
+  const getInitialData = React.useCallback(() => {
+    exportedStateRef.current = apiRef.current.exportState();
+
+    const rowIds = gridDataRowIdsSelector(apiRef);
+    const rowsLookup = gridRowsLookupSelector(apiRef);
+    const rows = rowIds.map((id) => rowsLookup[id]);
+
+    const columnFields = gridColumnFieldsSelector(apiRef);
+    const columnsLookup = gridColumnLookupSelector(apiRef);
+    const initialColumns = getInitialColumns(
+      columnFields,
+      columnsLookup,
+      props.getPivotDerivedColumns,
+    );
+
+    return { rows, columns: initialColumns };
+  }, [apiRef, props.getPivotDerivedColumns]);
+
+  const computePivotingState = React.useCallback(
+    ({ enabled, model: pivotModel }: Pick<GridPivotingState, 'enabled' | 'model'>) => {
+      if (enabled && pivotModel) {
+        const { rows, columns } = nonPivotDataRef.current || { rows: [], columns: new Map() };
+
+        return {
+          initialColumns: columns,
+          // TODO: fix getPivotedData called twice in controlled mode
+          propsOverrides: getPivotedData({
+            rows,
+            columns,
+            pivotModel,
+            apiRef: apiRef as RefObject<GridApiPremium>,
+            pivotingColDef: props.pivotingColDef,
+          }),
+        };
+      }
+
+      return undefined;
+    },
+    [apiRef, props.pivotingColDef],
+  );
+
+  useOnMount(() => {
+    if (!isPivotingAvailable || !isPivot) {
+      return undefined;
+    }
+
+    const isLoading = gridRowsLoadingSelector(apiRef) ?? false;
+
+    const runPivoting = () => {
+      nonPivotDataRef.current = getInitialData();
+      apiRef.current.setState((state) => {
+        const pivotingState = {
+          ...state.pivoting,
+          ...computePivotingState(state.pivoting),
+        };
+        return {
+          ...state,
+          pivoting: pivotingState,
+        };
+      });
+    };
+
+    if (!isLoading) {
+      runPivoting();
+      return undefined;
+    }
+
+    const unsubscribe = apiRef.current?.store.subscribe(() => {
+      const loading = gridRowsLoadingSelector(apiRef);
```

Comment:
Consider using the fresh state instead, as `apiRef.current.state` may occasionally lag behind the latest state.
```suggestion
    const unsubscribe = apiRef.current?.store.subscribe((state) => {
      const loading = gridRowsLoadingSelector({ current: { state }});
```

Reply:
I reverted this change, because it caused the test fail: https://app.circleci.com/jobs/github/mui/mui-x/489287